In [2]:
import pandas as pd
import numpy as np

# 데이터 프레임 생성
1) 100에서 200 사이의 값을 60개를 생성합니다(np.random.randint 이용)

2) 생성한 결과를 12,5 형태로 변환합니다.

3) 변환된 데이터를 이용하여 df 데이터 프레임을 생성합니다.

In [3]:
np.random.seed(1234) # 랜덤 -> 랜덤X (값고정)
df = pd.DataFrame(np.random.randint(100,200,60).reshape(12, 5))
df

0    1    2    3    4
0   147  183  138  153  176
1   124  115  149  123  126
2   130  143  130  126  158
3   192  169  180  173  147
4   150  176  137  134  138
5   167  111  100  175  180
6   103  102  119  112  165
7   175  181  114  171  160
8   146  128  181  187  113
9   196  112  169  195  131
10  189  184  145  116  141
11  172  156  170  156  186

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>147</td>
      <td>183</td>
      <td>138</td>
      <td>153</td>
      <td>176</td>
    </tr>
    <tr>
      <th>1</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
    </tr>
    <tr>
      <th>2</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
    </tr>
    <tr>
      <th>3</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
    </tr>
    <tr>
      <th>4</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
    </tr>
    <tr>
      <th>5</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
    </tr>
    <tr>
      <th>6</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
    </tr>
    <tr>
      <th>7</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
    </tr>
    <tr>
      <th>8</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
    </tr>
    <tr>
      <th>9</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
    </tr>
    <tr>
      <th>10</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
    </tr>
    <tr>
      <th>11</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
    </tr>
  </tbody>
</table>
</div>

# 컬럼명 변경
- 컬럼명을 ['P01', 'P02', 'P03', 'P04', 'P05']로 수정하세요.

In [4]:
# 코드를 입력해 주세요.
# columns 속성 이용
df.columns = ['P01', 'P02', 'P03', 'P04', 'P05']
df

P01  P02  P03  P04  P05
0   147  183  138  153  176
1   124  115  149  123  126
2   130  143  130  126  158
3   192  169  180  173  147
4   150  176  137  134  138
5   167  111  100  175  180
6   103  102  119  112  165
7   175  181  114  171  160
8   146  128  181  187  113
9   196  112  169  195  131
10  189  184  145  116  141
11  172  156  170  156  186

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>147</td>
      <td>183</td>
      <td>138</td>
      <td>153</td>
      <td>176</td>
    </tr>
    <tr>
      <th>1</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
    </tr>
    <tr>
      <th>2</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
    </tr>
    <tr>
      <th>3</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
    </tr>
    <tr>
      <th>4</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
    </tr>
    <tr>
      <th>5</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
    </tr>
    <tr>
      <th>6</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
    </tr>
    <tr>
      <th>7</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
    </tr>
    <tr>
      <th>8</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
    </tr>
    <tr>
      <th>9</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
    </tr>
    <tr>
      <th>10</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
    </tr>
    <tr>
      <th>11</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
    </tr>
  </tbody>
</table>
</div>

# 인덱스 변경
- 인덱스를 2020년 1월 부터 2020년 12월까지 마지막 요일로 설정하세요. 

In [5]:
# index 생성 
# date_range 함수 이용 (주기는 월(M)으로 설정)

date = pd.date_range('20200101','20201231', freq='M')
display(date)

DatetimeIndex(['2020-01-31', '2020-02-29', '2020-03-31', '2020-04-30',
               '2020-05-31', '2020-06-30', '2020-07-31', '2020-08-31',
               '2020-09-30', '2020-10-31', '2020-11-30', '2020-12-31'],
              dtype='datetime64[ns]', freq='M')

<p><strong>[출력 결과]</strong></p><pre>DatetimeIndex(['2020-01-31', '2020-02-29', '2020-03-31', '2020-04-30',
               '2020-05-31', '2020-06-30', '2020-07-31', '2020-08-31',
               '2020-09-30', '2020-10-31', '2020-11-30', '2020-12-31'],
              dtype='datetime64[ns]', freq='M')</pre>

In [6]:
# index 속성 이용하여 date 변수를 index로 지정하세요.
df.index = date
df

P01  P02  P03  P04  P05
2020-01-31  147  183  138  153  176
2020-02-29  124  115  149  123  126
2020-03-31  130  143  130  126  158
2020-04-30  192  169  180  173  147
2020-05-31  150  176  137  134  138
2020-06-30  167  111  100  175  180
2020-07-31  103  102  119  112  165
2020-08-31  175  181  114  171  160
2020-09-30  146  128  181  187  113
2020-10-31  196  112  169  195  131
2020-11-30  189  184  145  116  141
2020-12-31  172  156  170  156  186

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>147</td>
      <td>183</td>
      <td>138</td>
      <td>153</td>
      <td>176</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
    </tr>
    <tr>
      <th>2020-03-31</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
    </tr>
    <tr>
      <th>2020-04-30</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
    </tr>
    <tr>
      <th>2020-05-31</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
    </tr>
    <tr>
      <th>2020-06-30</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
    </tr>
    <tr>
      <th>2020-08-31</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
    </tr>
    <tr>
      <th>2020-09-30</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
    </tr>
    <tr>
      <th>2020-10-31</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
    </tr>
    <tr>
      <th>2020-11-30</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
    </tr>
    <tr>
      <th>2020-12-31</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
    </tr>
  </tbody>
</table>
</div>

# 데이터 선택
- 2020년 4월부터 6월까지 행을 출력하세요.

In [7]:
# 코드를 입력해 주세요.
df[3:6]

P01  P02  P03  P04  P05
2020-04-30  192  169  180  173  147
2020-05-31  150  176  137  134  138
2020-06-30  167  111  100  175  180

In [8]:
df.loc["2020-04-30":"2020-06-30", :] #df.loc(로케이션을 추출한다.)

P01  P02  P03  P04  P05
2020-04-30  192  169  180  173  147
2020-05-31  150  176  137  134  138
2020-06-30  167  111  100  175  180

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-04-30</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
    </tr>
    <tr>
      <th>2020-05-31</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
    </tr>
    <tr>
      <th>2020-06-30</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
    </tr>
  </tbody>
</table>
</div>

- 2020년 05월 31일의 P01 컬럼 ~ P03 컬럼을 출력하세요.

In [9]:
# 코드를 입력해 주세요.
df.loc["2020-05-31", '1':'3'] 

Series([], Name: 2020-05-31 00:00:00, dtype: int64)

<p><strong>[출력 결과]</strong></p><pre>P01    150
P02    176
P03    137
Name: 2020-05-31 00:00:00, dtype: int64</pre>

- 2020년 01월 31일의 P01 컬럼부터 P05 컬럼의 값을 0으로 설정하세요.

In [19]:
# 코드를 입력해 주세요.(여기 까지 하면 정말 잘한거다.)
  # hint. df.loc[행 인덱스, 열 인덱스] = 값

df.loc["2020-01-31", :] = 0 
df

P01  P02  P03  P04  P05    0
2020-01-31    0    0    0    0    0  0.0
2020-02-29  124  115  149  123  126  NaN
2020-03-31  130  143  130  126  158  NaN
2020-04-30  192  169  180  173  147  NaN
2020-05-31  150  176  137  134  138  NaN
2020-06-30  167  111  100  175  180  NaN
2020-07-31  103  102  119  112  165  NaN
2020-08-31  175  181  114  171  160  NaN
2020-09-30  146  128  181  187  113  NaN
2020-10-31  196  112  169  195  131  NaN
2020-11-30  189  184  145  116  141  NaN
2020-12-31  172  156  170  156  186  NaN

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
    </tr>
    <tr>
      <th>2020-03-31</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
    </tr>
    <tr>
      <th>2020-04-30</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
    </tr>
    <tr>
      <th>2020-05-31</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
    </tr>
    <tr>
      <th>2020-06-30</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
    </tr>
    <tr>
      <th>2020-08-31</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
    </tr>
    <tr>
      <th>2020-09-30</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
    </tr>
    <tr>
      <th>2020-10-31</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
    </tr>
    <tr>
      <th>2020-11-30</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
    </tr>
    <tr>
      <th>2020-12-31</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
    </tr>
  </tbody>
</table>
</div>

- 각 행의 합을 이용하여 Total 컬럼을 생성하세요
    - total = P01 + P02 + P03 + P04 + P05

In [11]:
# 코드를 입력해 주세요.
# hint. df[새로운컬럼명] = 값


<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
      <td>637</td>
    </tr>
    <tr>
      <th>2020-03-31</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
      <td>687</td>
    </tr>
    <tr>
      <th>2020-04-30</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
      <td>861</td>
    </tr>
    <tr>
      <th>2020-05-31</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
      <td>735</td>
    </tr>
    <tr>
      <th>2020-06-30</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
      <td>733</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
      <td>601</td>
    </tr>
    <tr>
      <th>2020-08-31</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
      <td>801</td>
    </tr>
    <tr>
      <th>2020-09-30</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
      <td>755</td>
    </tr>
    <tr>
      <th>2020-10-31</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
      <td>803</td>
    </tr>
    <tr>
      <th>2020-11-30</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
      <td>775</td>
    </tr>
    <tr>
      <th>2020-12-31</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
      <td>840</td>
    </tr>
  </tbody>
</table>
</div>

- Total 컬럼을 Total 컬럼의 총합 값으로 나누어 Total Rate 컬럼을 생성하세요.

In [12]:
# 코드를 입력해 주세요.


<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
      <th>Total Rate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
      <td>637</td>
      <td>7.74</td>
    </tr>
    <tr>
      <th>2020-03-31</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
      <td>687</td>
      <td>8.35</td>
    </tr>
    <tr>
      <th>2020-04-30</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
      <td>861</td>
      <td>10.46</td>
    </tr>
    <tr>
      <th>2020-05-31</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
      <td>735</td>
      <td>8.93</td>
    </tr>
    <tr>
      <th>2020-06-30</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
      <td>733</td>
      <td>8.91</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
      <td>601</td>
      <td>7.30</td>
    </tr>
    <tr>
      <th>2020-08-31</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
      <td>801</td>
      <td>9.74</td>
    </tr>
    <tr>
      <th>2020-09-30</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
      <td>755</td>
      <td>9.18</td>
    </tr>
    <tr>
      <th>2020-10-31</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
      <td>803</td>
      <td>9.76</td>
    </tr>
    <tr>
      <th>2020-11-30</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
      <td>775</td>
      <td>9.42</td>
    </tr>
    <tr>
      <th>2020-12-31</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
      <td>840</td>
      <td>10.21</td>
    </tr>
  </tbody>
</table>
</div>

- Total Rate 컬럼이 가장 낮은 3달을 출력하세요.

In [13]:
# 코드를 입력해 주세요.


<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
      <th>Total Rate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
      <td>601</td>
      <td>7.304327</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
      <td>637</td>
      <td>7.741857</td>
    </tr>
  </tbody>
</table>
</div>

- Total 값이 700 미만인 데이터를 출력하세요.

In [14]:
# 코드를 입력해 주세요.


<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
      <th>Total Rate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
      <td>637</td>
      <td>7.741857</td>
    </tr>
    <tr>
      <th>2020-03-31</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
      <td>687</td>
      <td>8.349538</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
      <td>601</td>
      <td>7.304327</td>
    </tr>
  </tbody>
</table>
</div>

- Total 값이 700 미만이면 Bad 700 이상이면 Good의 값을 가지는 state 컬럼을 생성하세요.

In [15]:
# 코드를 입력해 주세요.
  # hint. df.loc[조건, 컬럼명] = 값

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
      <th>Total Rate</th>
      <th>state</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-02-29</th>
      <td>124</td>
      <td>115</td>
      <td>149</td>
      <td>123</td>
      <td>126</td>
      <td>637</td>
      <td>7.741857</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-03-31</th>
      <td>130</td>
      <td>143</td>
      <td>130</td>
      <td>126</td>
      <td>158</td>
      <td>687</td>
      <td>8.349538</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-04-30</th>
      <td>192</td>
      <td>169</td>
      <td>180</td>
      <td>173</td>
      <td>147</td>
      <td>861</td>
      <td>10.464268</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-05-31</th>
      <td>150</td>
      <td>176</td>
      <td>137</td>
      <td>134</td>
      <td>138</td>
      <td>735</td>
      <td>8.932912</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-06-30</th>
      <td>167</td>
      <td>111</td>
      <td>100</td>
      <td>175</td>
      <td>180</td>
      <td>733</td>
      <td>8.908605</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-07-31</th>
      <td>103</td>
      <td>102</td>
      <td>119</td>
      <td>112</td>
      <td>165</td>
      <td>601</td>
      <td>7.304327</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-08-31</th>
      <td>175</td>
      <td>181</td>
      <td>114</td>
      <td>171</td>
      <td>160</td>
      <td>801</td>
      <td>9.735051</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-09-30</th>
      <td>146</td>
      <td>128</td>
      <td>181</td>
      <td>187</td>
      <td>113</td>
      <td>755</td>
      <td>9.175984</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-10-31</th>
      <td>196</td>
      <td>112</td>
      <td>169</td>
      <td>195</td>
      <td>131</td>
      <td>803</td>
      <td>9.759358</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-11-30</th>
      <td>189</td>
      <td>184</td>
      <td>145</td>
      <td>116</td>
      <td>141</td>
      <td>775</td>
      <td>9.419057</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-12-31</th>
      <td>172</td>
      <td>156</td>
      <td>170</td>
      <td>156</td>
      <td>186</td>
      <td>840</td>
      <td>10.209042</td>
      <td>Good</td>
    </tr>
  </tbody>
</table>
</div>

- state 컬럼의 Good과 Bad의 갯수를 계산하세요.

In [16]:
# 코드를 입력해 주세요.


<p><strong>[출력 결과]</strong></p><pre>Good    8
Bad     4
Name: state, dtype: int64</pre>

- 합계 행을 추가하세요.

In [17]:
# 코드를 입력해 주세요.


<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
      <th>Total Rate</th>
      <th>state</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31 00:00:00</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-02-29 00:00:00</th>
      <td>124.000000</td>
      <td>115.000000</td>
      <td>149.000000</td>
      <td>123.000000</td>
      <td>126.000000</td>
      <td>637.000000</td>
      <td>7.741857</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-03-31 00:00:00</th>
      <td>130.000000</td>
      <td>143.000000</td>
      <td>130.000000</td>
      <td>126.000000</td>
      <td>158.000000</td>
      <td>687.000000</td>
      <td>8.349538</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-04-30 00:00:00</th>
      <td>192.000000</td>
      <td>169.000000</td>
      <td>180.000000</td>
      <td>173.000000</td>
      <td>147.000000</td>
      <td>861.000000</td>
      <td>10.464268</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-05-31 00:00:00</th>
      <td>150.000000</td>
      <td>176.000000</td>
      <td>137.000000</td>
      <td>134.000000</td>
      <td>138.000000</td>
      <td>735.000000</td>
      <td>8.932912</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-06-30 00:00:00</th>
      <td>167.000000</td>
      <td>111.000000</td>
      <td>100.000000</td>
      <td>175.000000</td>
      <td>180.000000</td>
      <td>733.000000</td>
      <td>8.908605</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-07-31 00:00:00</th>
      <td>103.000000</td>
      <td>102.000000</td>
      <td>119.000000</td>
      <td>112.000000</td>
      <td>165.000000</td>
      <td>601.000000</td>
      <td>7.304327</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-08-31 00:00:00</th>
      <td>175.000000</td>
      <td>181.000000</td>
      <td>114.000000</td>
      <td>171.000000</td>
      <td>160.000000</td>
      <td>801.000000</td>
      <td>9.735051</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-09-30 00:00:00</th>
      <td>146.000000</td>
      <td>128.000000</td>
      <td>181.000000</td>
      <td>187.000000</td>
      <td>113.000000</td>
      <td>755.000000</td>
      <td>9.175984</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-10-31 00:00:00</th>
      <td>196.000000</td>
      <td>112.000000</td>
      <td>169.000000</td>
      <td>195.000000</td>
      <td>131.000000</td>
      <td>803.000000</td>
      <td>9.759358</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-11-30 00:00:00</th>
      <td>189.000000</td>
      <td>184.000000</td>
      <td>145.000000</td>
      <td>116.000000</td>
      <td>141.000000</td>
      <td>775.000000</td>
      <td>9.419057</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-12-31 00:00:00</th>
      <td>172.000000</td>
      <td>156.000000</td>
      <td>170.000000</td>
      <td>156.000000</td>
      <td>186.000000</td>
      <td>840.000000</td>
      <td>10.209042</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>합계</th>
      <td>1,744.000000</td>
      <td>1,577.000000</td>
      <td>1,594.000000</td>
      <td>1,668.000000</td>
      <td>1,645.000000</td>
      <td>8,228.000000</td>
      <td>nan</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

- Total 컬럼의 합계 값을 이용하여 P01부터 P05까지의 비율을 계산 후 합계 비율 행을 추가하세요.

In [18]:
# 코드를 입력해 주세요.


<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>P01</th>
      <th>P02</th>
      <th>P03</th>
      <th>P04</th>
      <th>P05</th>
      <th>Total</th>
      <th>Total Rate</th>
      <th>state</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-31 00:00:00</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-02-29 00:00:00</th>
      <td>124.000000</td>
      <td>115.000000</td>
      <td>149.000000</td>
      <td>123.000000</td>
      <td>126.000000</td>
      <td>637.000000</td>
      <td>7.741857</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-03-31 00:00:00</th>
      <td>130.000000</td>
      <td>143.000000</td>
      <td>130.000000</td>
      <td>126.000000</td>
      <td>158.000000</td>
      <td>687.000000</td>
      <td>8.349538</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-04-30 00:00:00</th>
      <td>192.000000</td>
      <td>169.000000</td>
      <td>180.000000</td>
      <td>173.000000</td>
      <td>147.000000</td>
      <td>861.000000</td>
      <td>10.464268</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-05-31 00:00:00</th>
      <td>150.000000</td>
      <td>176.000000</td>
      <td>137.000000</td>
      <td>134.000000</td>
      <td>138.000000</td>
      <td>735.000000</td>
      <td>8.932912</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-06-30 00:00:00</th>
      <td>167.000000</td>
      <td>111.000000</td>
      <td>100.000000</td>
      <td>175.000000</td>
      <td>180.000000</td>
      <td>733.000000</td>
      <td>8.908605</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-07-31 00:00:00</th>
      <td>103.000000</td>
      <td>102.000000</td>
      <td>119.000000</td>
      <td>112.000000</td>
      <td>165.000000</td>
      <td>601.000000</td>
      <td>7.304327</td>
      <td>Bad</td>
    </tr>
    <tr>
      <th>2020-08-31 00:00:00</th>
      <td>175.000000</td>
      <td>181.000000</td>
      <td>114.000000</td>
      <td>171.000000</td>
      <td>160.000000</td>
      <td>801.000000</td>
      <td>9.735051</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-09-30 00:00:00</th>
      <td>146.000000</td>
      <td>128.000000</td>
      <td>181.000000</td>
      <td>187.000000</td>
      <td>113.000000</td>
      <td>755.000000</td>
      <td>9.175984</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-10-31 00:00:00</th>
      <td>196.000000</td>
      <td>112.000000</td>
      <td>169.000000</td>
      <td>195.000000</td>
      <td>131.000000</td>
      <td>803.000000</td>
      <td>9.759358</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-11-30 00:00:00</th>
      <td>189.000000</td>
      <td>184.000000</td>
      <td>145.000000</td>
      <td>116.000000</td>
      <td>141.000000</td>
      <td>775.000000</td>
      <td>9.419057</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>2020-12-31 00:00:00</th>
      <td>172.000000</td>
      <td>156.000000</td>
      <td>170.000000</td>
      <td>156.000000</td>
      <td>186.000000</td>
      <td>840.000000</td>
      <td>10.209042</td>
      <td>Good</td>
    </tr>
    <tr>
      <th>합계</th>
      <td>1,744.000000</td>
      <td>1,577.000000</td>
      <td>1,594.000000</td>
      <td>1,668.000000</td>
      <td>1,645.000000</td>
      <td>8,228.000000</td>
      <td>nan</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>합계 비율</th>
      <td>0.211959</td>
      <td>0.191663</td>
      <td>0.193729</td>
      <td>0.202722</td>
      <td>0.199927</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>